test/unit-tests/expression/parse.test.js

```diff
@@ -2231,4 +2241,228 @@ describe('parse', function () {
 
     assert.strictEqual(mathClone.evaluate('2'), 2)
   })
+
+  describe('sources', function () {
+    it('adds sources for constants', function () {
+      const parsed = math.parse('4')
+      assert.deepStrictEqual(parsed.sources, [{ index: 0, text: '4' }])
+    })
+
+    it('adds sources for symbols', function () {
+      const parsed = math.parse('foo')
+      assert.deepStrictEqual(parsed.sources, [{ index: 0, text: 'foo' }])
+    })
+
+    it('adds sources for operators', function () {
+      const parsed = math.parse('1 + 2')
+      assert.deepStrictEqual(parsed.sources, [{ index: 2, text: '+' }])
+      assert.deepStrictEqual(parsed.args[0].sources, [{ index: 0, text: '1' }])
+      assert.deepStrictEqual(parsed.args[1].sources, [{ index: 4, text: '2' }])
+    })
+
+    it('adds sources for blocks', function () {
+      const parsed = math.parse('1 + 1; 2 + 2\n3 + 3')
+
+      // should have a source for each block delimiter
+      const expected = [
+        { index: 5, text: ';' },
+        { index: 12, text: '\n' }
+      ]
+
+      assert.deepStrictEqual(parsed.sources, expected)
+    })
+
+    it('adds sources for 1D matrices', function () {
+      const parsed = math.parse('[1, 2, 3]')
+
+      // should have a source for brackets and item delimiters
+      const expected = [
+        { index: 0, text: '[' },
+        { index: 2, text: ',' },
```

Comment:
If we parser a large matrix, I can imagine that listing all comma's impacts the performance. I think we should double check if this is an issue or not.

Reply:
Can do! Are you worried about performance during parsing or evaluation or somewhere else?

Reply:
Yes I was thinking about performance of parsing, there is a lot of extra objects created then. I think it cannot impact evaluation performance since the `sources` are not used during evaluation.

We can set up a small benchmark in the `/test/benchmark` folder to test if there is a performance impact or not, if you want I can do that.

Reply:
If you're willing to do that, I'm glad for the help, thank you! Otherwise, I'll tackle it after finishing everything else

Reply:
I'll do that 👍

Reply:
I've run the benchmark `test/benchmark/expression_parser.js` and compared the develop branch with this PR (you can try it out yourself to see the differences).

- The `evaluate` benchmark is just as fast in both cases, that is most important: about 1.7 million ops/sec
- The `parse` benchmark is slower, goes from 16000 ops/second to 11000 ops/sec. I think that is acceptable, and simply a consequence of the parse step having to do more work.

Shall we leave it at that?